velox/expression/Expr.cpp

```diff
@@ -259,6 +289,110 @@ bool isFlat(const BaseVector& vector) {
 
 } // namespace
 
+bool MutableRemainingRows::deselectNonErrorNulls(
+    const VectorPtr& result,
+    EvalCtx::ErrorVectorPtr& errors) {
+  const uint64_t* flatNulls = nullptr;
+  if (result->mayHaveNulls()) {
+    decoded_.get()->decode(*result, rows());
+    flatNulls = decoded_.get()->nulls();
+  }
+  // A null with no error deselects the row and clears a possible error. An
+  // error adds itself to argument errors.
+  if (context_.errors()) {
+    // There are new errors.
+    auto& remainingRows = mutableRows();
+    context_.ensureErrorsVectorSize(*context_.errorsPtr(), rows().end());
+    auto newErrors = context_.errors();
+    if (flatNulls) {
+      // There are both nulls and errors. Only a null with no error removes a
+      // row.
+      auto errorNulls = newErrors->rawNulls();
+      auto rowBits = mutableRows().asMutableRange().bits();
+      auto nwords = bits::nwords(rows().end());
+      for (auto i = 0; i < nwords; ++i) {
+        auto nullNoError = errorNulls ? flatNulls[i] | errorNulls[i] : 0;
+        rowBits[i] &= nullNoError;
+        if (errors) {
+          errors->mutableRawNulls()[i] &= nullNoError;
+        }
+      }
+      mutableRows().updateBounds();
+    }
+    context_.moveAppendErrors(errors);
+  } else {
+    // There are no new errors, a null deselects a row and clears an error.
+    if (flatNulls) {
+      if (errors) {
+        errors->addNulls(flatNulls, rows());
+      }
+      deselectNulls(flatNulls);
+    }
+  }
+  return rows().hasSelections();
+}
+
+template <typename EvalArg>
+bool Expr::evalArgsDefaultNulls(
```

Comment:
nit: I think the logic here can be simplified. Instead of modifying both context and argumentErrors inside deselectNonErrorNulls(), we keep track of all errors thrown in argumentErrors, then after evaluating all inputs, remainingRows would only consist of rows that either contain values or errors. So, we use remainingRows to copy only errors for those remaining rows by calling mergeOrThrowArgumentErrors() with remainingRows instead of original rows.

This way we dont have to keep track of updating nulls on argumentErrors and deselectNonErrorNulls() can be updated to only modify remainingRows.

```
template <typename EvalArg>
bool Expr::evalArgsDefaultNulls(
    MutableRemainingRows& rows,
    EvalArg evalArg,
    EvalCtx& context,
    VectorPtr& result) {
  EvalCtx::ErrorVectorPtr originalErrors;
  // After evaluating all inputs, this would contain the union of all errors
  // thrown during evaluation. Errors are overwritten if a previous evaluation
  // had thrown one at the same row index.
  EvalCtx::ErrorVectorPtr argumentErrors;
  // Store pre-existing errors locally and clear them from
  // 'context'. We distinguish between argument errors and
  // pre-existing (original) ones.
  if (context.errors()) {
    context.swapErrors(originalErrors);
  }
  inputValues_.resize(inputs_.size());
  {
    ScopedVarSetter throwErrors(
        context.mutableThrowOnError(), throwArgumentErrors(context));

    for (int32_t i = 0; i < inputs_.size(); ++i) {
      evalArg(i);
      context.moveAppendErrors(argumentErrors);
      if (!rows.deselectNonErrorNulls(inputValues_[i])) {
        break;
      }
    }
  }
  // Only merge the errors where none of the inputs evaluated to null.
  mergeOrThrowArgumentErrors(
      rows.rows(), originalErrors, argumentErrors, context);

  if (!rows.deselectErrors()) {
    releaseInputValues(context);
    setAllNulls(rows.originalRows(), context, result);
    return false;
  }

  return true;
}
```